### PR TITLE
Combine inferences into full pipeline

### PR DESCRIPTION
**Pull request by tzsombor95**
I implemented some classes that contains pre/postprocessing and inference functions for both yolo and gdrn model to combine into a full pipeline #12 .
The implementation is not the cleanest, but its a good starting point. I tried to use most of your existing methods.

### CODE REVIEW

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 1]
```diff
@@ -1,2 +1,63 @@
...
+
+    else:
+        for i, filename in enumerate(rgb_file_names):
+            image_names[i] = (image_names[i], None)
+
+    return image_names
+
+
+if __name__ == "__main__":
+    image_paths = get_image_list("../../../datasets/BOP_DATASETS/lmo/test/000001/rgb", "../../../datasets/BOP_DATASETS/lmo/test/000001/depth")
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:39]: Add
```
import os.path as osp
cur_dir = osp.dirname(osp.abspath(__file__))
PROJ_ROOT = osp.normpath(osp.join(cur_dir, "../../.."))
sys.path.insert(0, PROJ_ROOT)
```
and access the paths like
```
osp.join(PROJ_ROOT, "datasets/xxx")
```
to make it runnable everywhere.

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 2]
```diff
@@ -1,2 +1,63 @@
...
+    yolo_predictor = YoloPredictor(
+                       exp_name="yolox-x",
+                       config_file_path="../../../configs/yolox/bop_pbr/yolox_x_640_augCozyAAEhsv_ranger_30_epochs_lmo_pbr_lmo_bop_test.py",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:42]: same as above

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 3]
```diff
@@ -1,2 +1,63 @@
...
+                       ckpt_file_path="../../../output/yolox/bop_pbr/yolox_x_640_augCozyAAEhsv_ranger_30_epochs_lmo_pbr_lmo_bop_test/model_final.pth",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:43]: same as above

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 4]
```diff
@@ -1,2 +1,63 @@
...
+                       fuse=True,
+                       fp16=False
+                     )
+    gdrn_predictor = GdrnPredictor(
+        config_file_path="../../../configs/gdrn/lmo_pbr/convnext_a6_AugCosyAAEGray_BG05_mlL1_DMask_amodalClipBox_classAware_lmo.py",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:48]: same as above

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 5]
```diff
@@ -1,2 +1,63 @@
...
+        ckpt_file_path="../../../output/gdrn/lmo_pbr/convnext_a6_AugCosyAAEGray_BG05_mlL1_DMask_amodalClipBox_classAware_lmo/model_final.pth",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:49]: same as above

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 6]
```diff
@@ -1,2 +1,63 @@
...
+        camera_json_path="../../../datasets/BOP_DATASETS/lmo/camera.json",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:50]: same as above

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 7]
```diff
@@ -1,2 +1,63 @@
...
+        path_to_obj_models="../../../datasets/BOP_DATASETS/lmo/models"
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:51]: same as above

File: core/gdrn_modeling/demo/predictor_gdrn.py [hunk 8]
```diff
@@ -1,2 +1,626 @@
...
+    GDRN_no_region,
+    GDRN_cls,
+    GDRN_cls2reg,
+    GDRN_double_mask,
+    GDRN_Dstream_double_mask,
+)  # noqa
+
+class GdrnPredictor():
+    def __init__(self,
+                 config_file_path="../../../configs/gdrn/lmo_pbr/convnext_a6_AugCosyAAEGray_BG05_mlL1_DMask_amodalClipBox_classAware_lmo.py",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/predictor_gdrn.py:41]: same as above

File: core/gdrn_modeling/demo/predictor_gdrn.py [hunk 9]
```diff
@@ -1,2 +1,626 @@
...
+                 ckpt_file_path="../../../output/gdrn/lmo_pbr/convnext_a6_AugCosyAAEGray_BG05_mlL1_DMask_amodalClipBox_classAware_lmo/model_final.pth",
+                 camera_json_path="../../../datasets/BOP_DATASETS/lmo/camera.json",
+                 path_to_obj_models="../../../datasets/BOP_DATASETS/lmo/models"
```
Comment (wangg12) [inline core/gdrn_modeling/demo/predictor_gdrn.py:44]: same as above

File: core/gdrn_modeling/demo/predictor_yolo.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+from detectron2.evaluation import inference_context
+from setproctitle import setproctitle
+from types import SimpleNamespace
+from contextlib import ExitStack
+from loguru import logger
+
+class YoloPredictor():
+
+    def __init__(self, exp_name="yolox-x",
+                       config_file_path="../../../configs/yolox/bop_pbr/yolox_x_640_augCozyAAEhsv_ranger_30_epochs_lmo_pbr_lmo_bop_test.py",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/predictor_yolo.py:26]: same as above

File: core/gdrn_modeling/demo/predictor_yolo.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                                  keep_single_instance=True)
+
+        return outputs
+
+if __name__ == "__main__":
+    #little test example to visualize yolo model output
+
+    predictor = YoloPredictor(
+                       exp_name="yolox-x",
+                       config_file_path="../../../configs/yolox/bop_pbr/yolox_x_640_augCozyAAEhsv_ranger_30_epochs_lmo_pbr_lmo_bop_test.py",
```
Comment (wangg12) [inline core/gdrn_modeling/demo/predictor_yolo.py:172]: same as above

File: core/gdrn_modeling/demo/predictor_yolo.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+                       ckpt_file_path="../../../output/yolox/bop_pbr/yolox_x_640_augCozyAAEhsv_ranger_30_epochs_lmo_pbr_lmo_bop_test/model_final.pth",
+                       fuse=True,
+                       fp16=False
+                       )
+    img_path = "../../../datasets/BOP_DATASETS/lmo/test/000001/rgb/000000.jpg"
```
Comment (wangg12) [inline core/gdrn_modeling/demo/predictor_yolo.py:177]: same as above

**Comment by tzsombor95** [comment]
I fixed the paths.

File: core/gdrn_modeling/demo/demo_gdrn.py [hunk 13]
```diff
@@ -1,2 +1,70 @@
 # inference with detector, gdrn, and refiner
-# TODO
+import os.path as osp
+import sys
+cur_dir = osp.dirname(osp.abspath(__file__))
+PROJ_ROOT = osp.normpath(osp.join(cur_dir, "../../.."))
+sys.path.insert(0, PROJ_ROOT)
+
+
+from predictor_yolo import YoloPredictor
```
Comment (wangg12) [inline core/gdrn_modeling/demo/demo_gdrn.py:9]: I think here should be 
```
from core.gdrn_modeling.demo.predictor_yolo import YoloPredictor
from core.gdrn_modeling.demo.predictor_gdrn import GdrnPredictor
```

**Comment by wangg12** [comment]
@shanice-l Please review and test the MR when you have time.

**Comment by shanice-l** [comment]
I'll review it after the ICCV deadline (3.8).

**Comment by shailabhsuman** [comment]
I tired the PR on YCB dataset with path changes. prediction_yolo is fine but prediction gdrn is not. it classifies the data wrongly at inference of gdrn. Just my 5 cents to write which dataset it is tested with and works correctly. Looking into what could be wrong but this PR is a great work on top of already the SOTA for pose estimation

**Comment by gonalo99** [comment]
I also tried this code in a custom dataset and confirm what @shailabhsuman said, the pose results from predictor_gdrn are not the same as the ones obtained during test time.

**Comment by Martin9797** [comment]
After trying with both the tless dataset and a custom dataset (where this tool was a great help), I also found that while yolox is accurate, gdrn prediction and poses are most likely off. Has anyone found a fix yet?

**Comment by Martin9797** [comment]
I am having another issue with this; when using it on a picture with multiple objects of the same class in it, it only creates a box and poses for one (likely highest scoring one) of the objects in the picture. 
The trained Network does however find all the objects and poses when evaluated on the same image using the bop_toolkit visualizer for estimated poses.
Does anyone have an idea where the issue might be?

**Comment by freLorbeer** [comment]
Thanks for your code! 
I am wondering if this code really uses the depth refinement module (Coupled Iterative Refinement for 6D Multi-Object Pose Estimation)

As far as I see it uses opencv PNP solver, right?
Is there already an inference pipeline implementation that uses the depth refinement module ?

**Comment by shanice-l** [comment]
@freLorbeer  Plz check out the depth_refine branch for the code.

**Pull request merged**